## Data Cleaning

In [1]:
import pandas as pd
import numpy as np

#import your edited csv:
df = pd.read_csv("fuel_prices_cleaned.csv", index_col=0)
df

,Name,Super95,Diesel,ZIP,State,District,City,Address,Date,coordinates_X,coordinates_Y
0,Lagerhaus Tankstelle,1.799,1.819,9640,Kärnten,Hermagor,Kötschach,Kötschach 153,2022-08-13 17:57:00,13.0032407,46.672118
1,Turmöl Quick,1.769,1.780,2091,Niederösterreich,Horn,Langau,Langau 195,2022-08-13 17:57:00,15.7170359,48.829642
2,Genol Lagerhaus Tankstelle,0.000,1.799,3851,Niederösterreich,Waidhofen a. d. Thaya,Kautzen,Illmauerstraße 1,2022-08-13 17:57:00,15.240267,48.929879
3,Turmöl Quick,1.714,1.734,4048,Oberösterreich,Urfahr Umgebung,Puchenau,Hammerschmiede 3,2022-08-13 17:57:00,14.2412365,48.312442
4,Avia Xpress (Automaten-Tankstelle),1.704,1.712,4563,Oberösterreich,Kirchdorf,Micheldorf,Weinzierler Brücke 72,2022-08-13 17:57:00,14.129979,47.894056
...,...,...,...,...,...,...,...,...,...,...,...
25664,Turmöl,0.000,1.898,4910,Oberösterreich,Ried,Ried im Innkreis,Braunauerstraße 10,2022-09-29 07:06:00,13.46058,48.205890
25665,Turbosprit,1.628,1.858,9130,Kärnten,Klagenfurt Land,Poggersdorf,Landesstraße 2,2022-09-29 07:05:00,14.449627,46.641242
25666,Turmöl,1.614,1.874,1230,Wien,Liesing,Wien,Altmannsdorferstraße 150,2022-09-29 07:07:00,16.31935,48.154766
25667,Turmöl Quick,1.594,1.864,3150,Niederösterreich,St.Pölten Land,Wilhelmsburg,Lilienfelderstraße 37,2022-09-29 07:06:00,15.60052,48.100190


In [2]:
#get rid of 0.0 - how?

#dropping rows where value=0?
df_c = df
df_c = df_c[df_c["Diesel"] != 0]
df_c = df_c[df_c["Super95"] != 0]

#dataframes can be used to show a nicely formatted table:
pd.DataFrame({"is": df_c[["Super95", "Diesel"]].mean(), "actual": [1.713895,1.85288]})

,is,actual
Super95,1.715234,1.713895
Diesel,1.985328,1.852880


better, but still wrong! - Why?

In [3]:
# Dropping rows changes your data! A price of 0.0 means missing values -> replace by NaN!
df = df.replace(0.0, np.nan)

# Statistics
pd.DataFrame({"is": df[["Super95", "Diesel"]].mean(), "actual": [1.713895,1.85288]})

,is,actual
Super95,1.714699,1.713895
Diesel,1.942319,1.852880


Better, but still wrong. There's still a problem with the data.

In [4]:
df.describe()

,Super95,Diesel,ZIP,coordinates_Y
count,18616.000000,19826.000000,25669.000000,25669.000000
mean,1.714699,1.942319,5160.815848,47.783986
std,0.123280,12.593193,2572.876362,0.610747
min,1.469000,1.469000,1010.000000,46.489800
25%,1.679000,1.789000,3180.000000,47.272684
50%,1.709000,1.865000,4780.000000,47.937198
75%,1.745000,1.913000,7143.000000,48.245661
max,16.630000,1775.000000,9990.000000,48.943722


look at max. values: is a price of 16.63 or 1775 Euro plausible?
### 📝 Tasks for you:

In [5]:
#Show prices (for both fuel types) which are higher than 3 Euros:
df[(df['Super95']>3) | (df['Diesel']>3)]

,Name,Super95,Diesel,ZIP,State,District,City,Address,Date,coordinates_X,coordinates_Y
243,Lagerhaus,1.744,1775.0,9400,Kärnten,Wolfsberg,Wolfsberg,Tanglstraße 2,2022-08-13 17:57:00,14.8407096,46.844043
344,Eni Austria,16.630,NaN,8934,Steiermark,Liezen,Altenmarkt bei St.Gallen,Altenmarkt Nr.81,2022-08-13 17:57:00,14.6540434,47.719330


discuss, why are these wrong? What possibly happened?

In [6]:
#delete rows - find an other way based on your discussion before and replace following line:
df.drop([243,344])

,Name,Super95,Diesel,ZIP,State,District,City,Address,Date,coordinates_X,coordinates_Y
0,Lagerhaus Tankstelle,1.799,1.819,9640,Kärnten,Hermagor,Kötschach,Kötschach 153,2022-08-13 17:57:00,13.0032407,46.672118
1,Turmöl Quick,1.769,1.780,2091,Niederösterreich,Horn,Langau,Langau 195,2022-08-13 17:57:00,15.7170359,48.829642
2,Genol Lagerhaus Tankstelle,NaN,1.799,3851,Niederösterreich,Waidhofen a. d. Thaya,Kautzen,Illmauerstraße 1,2022-08-13 17:57:00,15.240267,48.929879
3,Turmöl Quick,1.714,1.734,4048,Oberösterreich,Urfahr Umgebung,Puchenau,Hammerschmiede 3,2022-08-13 17:57:00,14.2412365,48.312442
4,Avia Xpress (Automaten-Tankstelle),1.704,1.712,4563,Oberösterreich,Kirchdorf,Micheldorf,Weinzierler Brücke 72,2022-08-13 17:57:00,14.129979,47.894056
...,...,...,...,...,...,...,...,...,...,...,...
25664,Turmöl,NaN,1.898,4910,Oberösterreich,Ried,Ried im Innkreis,Braunauerstraße 10,2022-09-29 07:06:00,13.46058,48.205890
25665,Turbosprit,1.628,1.858,9130,Kärnten,Klagenfurt Land,Poggersdorf,Landesstraße 2,2022-09-29 07:05:00,14.449627,46.641242
25666,Turmöl,1.614,1.874,1230,Wien,Liesing,Wien,Altmannsdorferstraße 150,2022-09-29 07:07:00,16.31935,48.154766
25667,Turmöl Quick,1.594,1.864,3150,Niederösterreich,St.Pölten Land,Wilhelmsburg,Lilienfelderstraße 37,2022-09-29 07:06:00,15.60052,48.100190


In [7]:
df.describe()

,Super95,Diesel,ZIP,coordinates_Y
count,18616.000000,19826.000000,25669.000000,25669.000000
mean,1.714699,1.942319,5160.815848,47.783986
std,0.123280,12.593193,2572.876362,0.610747
min,1.469000,1.469000,1010.000000,46.489800
25%,1.679000,1.789000,3180.000000,47.272684
50%,1.709000,1.865000,4780.000000,47.937198
75%,1.745000,1.913000,7143.000000,48.245661
max,16.630000,1775.000000,9990.000000,48.943722


### Extracting more Infos - derived data

In [9]:
# Find all full-self-service gas stations
df['Automat'] = df['Name'].str.contains('utomat')

#create new column with on-premise gas stations (Hofer, Lagerhaus)
# 1. Define the conditions
conditions = [
    df['Address'].str.contains('Hofer Pp.', na=False),
    df['Name'].str.contains('agerhaus', na=False),
    df['Address'].str.contains('Lagerhaus ', na=False),
    df['Name'].str.contains('Genol', na=False)
]

# 2. Define the corresponding values
choices = ['Hofer', 'Lagerhaus', 'Lagerhaus', 'Lagerhaus']

# 3. Create the new column (assign 'Unknown' or None if no match)
df['OnPrem'] = np.select(conditions, choices, default='Unknown')

### 📝 Tasks for you:

In [ ]:
#how many rows does our dataset have?
dataset_rows = #25669
#how many values are there for Diesel prices?
dataset_values_diesel = #19826
#how many values are there for Super95 prices?
dataset_values_super95 = #18616
#what is the lowest Diesel price?
lowest_diesel = #1.469
#what is the highest Super95 price?
highest_super95 = #2.199
#which date range does the dataset cover?
dataset_daterange = ["", ""]

#How many automated gas stations are there in the dataset?
nr_automat_stations = #935